File: wordwrap/src/main/scala/WordWrap.scala
/*
 * From https://gist.github.com/thunklife/4cbb85d6a011a5a12638deb2875730b3
 *
 * Write a function, wrap that takes two arguments, a string, and a column number.
 * The function returns the string, but with line breaks inserted in the right places
 * based column number passed to the function.
 *
 * Make sure that no line is longer than the column number and only break lines at word boundaries.
 *
 * Like a word processor, break the line by replacing the last space in a line with a newline.
 */

object WordWrap extends App {

  /*
  def break(pos: Int): String = str.substring (0, pos) + '\n' + wrap(str.substring (pos + 1), col)

  if (str.length <= col)
    str
  else {
    val breakPos = str.lastIndexWhere(char => char.isWhitespace, col)
    if (breakPos < 0) { // no suitable whitespace character to break on
      // find the next whitespace character and use that
      val firstAvailableBreak = str.indexWhere(char => char.isWhitespace, col + 1)
      if (firstAvailableBreak < 0) str else break(firstAvailableBreak)
    }
    else
      break(breakPos)
  }
  */


  def wrap(str: String, col: Int): String = {
    // General idea:
    //   - if string length is <= col, no breaking is necessary; otherwise:
    //   - find last word break at position <= column number
    //   - return [chars before the break] + '\n' + recursive call of chars after the break

    def break(pos: Int): String = pos match {
      case -1 => str
      case _  => str.substring(0, pos) + '\n' + wrap(str.substring(pos + 1), col)
    }

    val breakPos = {
      if (str.length <= col)
        -1
      else {
        val p = str.lastIndexWhere(char => char.isWhitespace, col)
        if (p >= 0) p else str.indexWhere(char => char.isWhitespace, col + 1)
      }
    }

    break(breakPos)
  }

  // TESTING

  val testString = "Write a function, wrap that takes two arguments, a string, and a column number. The function returns the string, but with line breaks inserted in the right places based column number passed to the function. Make sure that no line is longer than the column number and only break lines at word boundaries."

  println(testString)
  println(wrap(testString, 79))

  println(wrap("abcdefghjik lmnopqrstuvwxyz", 10)) // hmm, should this break or not?

}

File: wordwrap/wordwrap.hs
import Data.Char
import Data.List

findBreakPosition :: [Char] -> Int -> Int
findBreakPosition [] col = 0
findBreakPosition str col
    | length str <= col = -1
    | otherwise =
        let breaks = takeWhile (<= col) $ findIndices isSpace str
            trivial = last breaks  -- TODO may be empty list, pattern match?
        in trivial
    -- TODO
    -- If length of str <= col, return -1 or other sentinel?
    -- Otherwise find last index of whitespace at of before cols

doBreak :: [Char] -> Int -> [Char]
doBreak str (-1) = str
doBreak str pos =
    let parts = splitAt pos str
    in (fst parts) ++ "\n" ++ wrap (drop 1 $ snd parts) pos

wrap :: [Char] -> Int -> [Char]
wrap str col = doBreak str $ findBreakPosition str col
